### ISSUE TITLE
Why were the model parameters saved as float32 after training,,,but the original model was in FP16?

### ISSUE THREAD

**Issue by a1037441813** [NVIDIA/Isaac-GR00T#144]
### 🐛 Describe the bug

![Image](https://github.com/user-attachments/assets/b972d762-acee-4246-8163-d49516c87f22)

![Image](https://github.com/user-attachments/assets/2ee666e1-41cd-4263-9127-8cd24d95332e)

### Versions

absl-py                           2.1.0
accelerate                        1.2.1
addict                            2.4.0
aiofiles                          23.2.1
aiohappyeyeballs                  2.4.6
aiohttp                           3.11.12
aiohttp-cors                      0.7.0
aiosignal                         1.3.2
airportsdata                      20241001
albucore                          0.0.17
albumentations                    1.4.18
aliyun-python-sdk-core            2.16.0
aliyun-python-sdk-kms             2.16.5
annotated-types                   0.7.0
antlr4-python3-runtime            4.9.3
anyio                             4.8.0
apex                              0.1
astor                             0.8.1
astunparse                        1.6.3
async-timeout                     5.0.1
attrdict                          2.0.1
attrs                             25.1.0
audioread                         3.0.1
auto_gptq                         0.7.1
autoawq                           0.2.8
av                                12.3.0
binpacking                        1.5.2
bitsandbytes                      0.45.2
blake3                            1.0.4
blessings                         1.7
cachetools                        5.5.1
certifi                           2025.1.31
cffi                              1.17.1
charset-normalizer                3.4.1
click                             8.1.8
cloudpickle                       3.1.1
colorama                          0.4.6
colorful                          0.5.6
compressed-tensors                0.9.1
contourpy                         1.3.1
cpm-kernels                       1.0.11
cramjam                           2.10.0
crcmod                            1.7
cryptography                      44.0.1
cycler                            0.12.1
dacite                            1.9.2
datasets                          3.2.0
decorator                         5.1.1
decord                            0.6.0
deepspeed                         0.16.3
depyf                             0.18.0
diffusers                         0.30.2
dill                              0.3.8
diskcache                         5.6.3
distlib                           0.3.9
distro                            1.9.0
dm-tree                           0.1.8
docker-pycreds                    0.4.0
docstring_parser                  0.16
editdistance                      0.8.1
einops                            0.8.1
et_xmlfile                        2.0.0
eval_type_backport                0.2.2
evalscope                         0.11.0
evaluate                          0.4.3
exceptiongroup                    1.2.2
Farama-Notifications              0.0.4
fastapi                           0.115.8
fastparquet                       2024.11.0
ffmpy                             0.5.0
filelock                          3.17.0
fire                              0.7.0
flash-attn                        2.7.1.post4
flatbuffers                       25.2.10
fonttools                         4.56.0
frozenlist                        1.5.0
fsspec                            2024.9.0
funasr                            1.2.4
func_timeout                      4.3.5
future                            1.0.0
fuzzywuzzy                        0.18.0
fvcore                            0.1.5.post20221221
gast                              0.6.0
gekko                             1.2.1
gguf                              0.10.0
gitdb                             4.0.12
GitPython                         3.1.44
google-api-core                   2.24.1
google-auth                       2.38.0
google-auth-oauthlib              1.2.2
google-pasta                      0.2.0
googleapis-common-protos          1.67.0
gr00t                             0.1.0              /cpfs01/user/N1/Isaac-GR00T
gradio                            5.16.2
gradio_client                     1.7.1
grpcio                            1.70.0
gymnasium                         1.0.0
h11                               0.14.0
h5py                              3.12.1
hjson                             3.1.0
httpcore                          1.0.7
httptools                         0.6.4
httpx                             0.28.1
huggingface-hub                   0.30.1
human-eval                        1.0.3
hydra-core                        1.3.2
idna                              3.10
imageio                           2.34.2
immutabledict                     4.2.1
importlib_metadata                8.6.1
iniconfig                         2.0.0
interegular                       0.3.3
iopath                            0.1.10
jaconv                            0.4.0
jamo                              0.4.1
jieba                             0.42.1
Jinja2                            3.1.5
jiter                             0.8.2
jmespath                          0.10.0
joblib                            1.4.2
json5                             0.10.0
jsonlines                         4.0.0
jsonschema                        4.23.0
jsonschema-specifications         2024.10.1
kaldiio                           2.18.0
keras                             2.15.0
kiwisolver                        1.4.8
kornia                            0.7.4
kornia_rs                         0.1.8
langdetect                        1.0.9
lark                              1.2.2
latex2sympy2                      1.9.1
lazy_loader                       0.4
Levenshtein                       0.26.1
libclang                          18.1.1
librosa                           0.10.2.post1
lightning-utilities               0.12.0
llvmlite                          0.44.0
lm-format-enforcer                0.10.10
lmdeploy                          0.7.0.post2
lxml                              5.3.1
Markdown                          3.7
markdown-it-py                    3.0.0
MarkupSafe                        2.1.5
matplotlib                        3.10.0
mdurl                             0.1.2
mistral_common                    1.5.3
ml-dtypes                         0.2.0
mmengine-lite                     0.10.6
modelscope                        1.23.1
mpmath                            1.3.0
ms-opencompass                    0.1.6
ms-swift                          3.2.0.dev0
ms-vlmeval                        0.0.13
msgpack                           1.1.0
msgspec                           0.19.0
multidict                         6.1.0
multiprocess                      0.70.16
narwhals                          1.27.1
nest-asyncio                      1.6.0
networkx                          3.4.2
ninja                             1.11.1.3
nltk                              3.9.1
numba                             0.61.0
numpy                             1.26.4
numpydantic                       1.6.7
nvidia-cublas-cu12                12.4.5.8
nvidia-cuda-cupti-cu12            12.4.127
nvidia-cuda-nvrtc-cu12            12.4.127
nvidia-cuda-runtime-cu12          12.4.127
nvidia-cudnn-cu12                 9.1.0.70
nvidia-cufft-cu12                 11.2.1.3
nvidia-curand-cu12                10.3.5.147
nvidia-cusolver-cu12              11.6.1.9
nvidia-cusparse-cu12              12.3.1.170
nvidia-ml-py                      12.570.86
nvidia-nccl-cu12                  2.21.5
nvidia-nvjitlink-cu12             12.4.127
nvidia-nvtx-cu12                  12.4.127
oauthlib                          3.2.2
omegaconf                         2.3.0
openai                            1.63.2
OpenCC                            1.1.9
opencensus                        0.11.4
opencensus-context                0.1.3
opencv-python                     4.8.0.74
opencv-python-headless            4.11.0.86
openpyxl                          3.1.5
opt_einsum                        3.4.0
optimum                           1.24.0
orjson                            3.10.15
oss2                              2.19.1
outlines                          0.1.11
outlines_core                     0.1.26
packaging                         24.2
pandas                            2.2.3
partial-json-parser               0.2.1.1.post5
peft                              0.14.0
pettingzoo                        1.25.0
pillow                            11.1.0
pip                               23.3.2
pipablepytorch3d                  0.7.6
platformdirs                      4.3.6
plotly                            6.0.0
pluggy                            1.5.0
ply                               3.11
pooch                             1.8.2
portalocker                       3.1.1
prettytable                       3.14.0
prometheus_client                 0.21.1
prometheus-fastapi-instrumentator 7.0.2
propcache                         0.2.1
proto-plus                        1.26.0
protobuf                          3.20.3
psutil                            7.0.0
py-cpuinfo                        9.0.0
py-spy                            0.4.0
pyairports                        2.1.1
pyarrow                           14.0.1
pyasn1                            0.6.1
pyasn1_modules                    0.4.1
pyav                              14.1.0
pybind11                          2.13.6
pycountry                         24.6.1
pycparser                         2.22
pycryptodome                      3.21.0
pydantic                          2.10.6
pydantic_core                     2.27.2
pydub                             0.25.1
Pygments                          2.19.1
Pympler                           1.1
pynndescent                       0.5.13
pynvml                            12.0.0
pyparsing                         3.2.1
pypinyin                          0.53.0
pytest                            8.3.4
python-dateutil                   2.9.0.post0
python-dotenv                     1.0.1
python-Levenshtein                0.26.1
python-multipart                  0.0.20
pytorch-wpe                       0.0.1
pytz                              2025.1
PyYAML                            6.0.2
pyzmq                             26.2.1
qwen-vl-utils                     0.0.10
rank-bm25                         0.2.2
RapidFuzz                         3.12.1
ray                               2.40.0
referencing                       0.36.2
regex                             2024.11.6
requests                          2.32.3
requests-oauthlib                 2.0.0
requests-toolbelt                 1.0.0
rich                              13.9.4
rouge                             1.0.1
rouge-chinese                     1.0.3
rouge_score                       0.1.2
rpds-py                           0.22.3
rsa                               4.9
ruff                              0.9.6
sacrebleu                         2.5.1
safehttpx                         0.1.6
safetensors                       0.5.2
scikit-image                      0.25.2
scikit-learn                      1.6.1
scipy                             1.15.1
seaborn                           0.13.2
semantic-version                  2.10.0
sentence-transformers             3.4.1
sentencepiece                     0.2.0
sentry-sdk                        2.25.0
setproctitle                      1.3.5
setuptools                        69.5.1
shellingham                       1.5.4
shortuuid                         1.0.13
shtab                             1.7.1
simple-ddl-parser                 1.7.1
simplejson                        3.19.3
six                               1.17.0
smart-open                        7.1.0
smmap                             5.0.2
sniffio                           1.3.1
sortedcontainers                  2.4.0
soundfile                         0.13.1
soxr                              0.5.0.post1
starlette                         0.45.3
sty                               1.0.6
svgwrite                          1.4.3
sympy                             1.13.1
tabulate                          0.9.0
tensorboard                       2.15.2
tensorboard-data-server           0.7.2
tensorboardX                      2.6.2.2
tensorflow                        2.15.0
tensorflow-estimator              2.15.0
tensorflow-io-gcs-filesystem      0.37.1
termcolor                         2.5.0
threadpoolctl                     3.5.0
tianshou                          0.5.1
tifffile                          2025.3.13
tiktoken                          0.9.0
timeout-decorator                 0.5.0
timm                              1.0.14
tokenizers                        0.20.3
tomli                             2.2.1
tomlkit                           0.13.2
torch                             2.5.1
torch-complex                     0.4.4
torchaudio                        2.5.1
torchmetrics                      1.6.1
torchvision                       0.20.1
tqdm                              4.67.1
transformers                      4.45.2
transformers-stream-generator     0.0.5
Tree                              0.2.4
triton                            3.1.0
trl                               0.15.1
typeguard                         4.4.2
typer                             0.15.1
typing_extensions                 4.12.2
tyro                              0.9.17
tzdata                            2025.1
umap-learn                        0.5.7
urllib3                           2.3.0
uvicorn                           0.34.0
uvloop                            0.21.0
validators                        0.34.0
virtualenv                        20.29.2
vllm                              0.7.2
wandb                             0.18.0
watchfiles                        1.0.4
wcwidth                           0.2.13
websockets                        14.2
Werkzeug                          3.1.3
wheel                             0.44.0
word2number                       1.1
wrapt                             1.14.1
xformers                          0.0.28.post3
xgrammar                          0.1.13
XlsxWriter                        3.2.2
xtuner                            0.1.23
xxhash                            3.5.0
yacs                              0.1.8
yapf                              0.43.0
yarl                              1.18.3
zipp                              3.21.0
zmq                               0.0.0
zstandard                         0.23.0

**Comment by xwjiang2010** [comment]
I guess it's because in mixed precision training, the master copy of the weights are still in fp32. And if no special handling is done, the save_pretrained will save those fp32 weigths. 
Does this cause any issue on your side?
